### DISCUSSION TITLE
Send a Request/Response between modules without circular dependency

### DISCUSSION THREAD

**Discussion by hmik** [spring-projects/spring-modulith#698] [Q&A]
This is my workflow:
Module A:
Controler gets a Request, passes the data to the service, the Service retrieves the data from the Repository and pubilshes an Event "GenerateDocFile"

Module B: 
Listens for the Event, extracts the data, calls the Service that generates the Doc file.

The Doc file has to be returned to Module A so it can be returned to controller and downloaded

I tried 2 approaches:
1) Module B sets up the Doc file in the event's CompletableFuture, Module A waits for future to be filled and reads the data in it
2) Module B publishes new even "DocGenerated", Module A listens for it and extracts data from it

While both approaches work, 
the modules.verify() tets always fails because it detects a cycle.

Is there a way to set this up so the test will be successful?

**Comment by werlo** [comment]
@hmik did you find any solution?? i am stuck too.

**Reply by hmik** [reply]
I don't know if it's the correct way, but adding allowedDependencies to the package-info of Module A stopped breaking the test

**Comment by jfoder** [comment]
It seems like module A **depends** on module B. I think that direct service call from module A to module B would be a much better approach to retrieve the doc file. Your event is actually a command, the result of which is necessary for module A to process the request. Calling the service directly looks much more natural and does not cause a cyclic dependency
